// convex/dist/esm/cli/lib/components/definition/directoryStructure.js
"use strict";
import path from "path";
import { logError } from "../../../../bundler/context.js";
import { DEFINITION_FILENAME, ROOT_DEFINITION_FILENAME } from "../constants.js";
export function qualifiedDefinitionPath(directory, workingDir = ".") {
  const definitionPath = path.relative(workingDir, directory.definitionPath);
  return `./${definitionPath}`;
}
export function isComponentDirectory(ctx, directory, isRoot) {
  if (!ctx.fs.exists(directory)) {
    return { kind: "err", why: `Directory doesn't exist` };
  }
  const dirStat = ctx.fs.stat(directory);
  if (!dirStat.isDirectory()) {
    return { kind: "err", why: `Not a directory` };
  }
  const filename = isRoot ? ROOT_DEFINITION_FILENAME : DEFINITION_FILENAME;
  const definitionPath = path.resolve(path.join(directory, filename));
  if (!ctx.fs.exists(definitionPath)) {
    return {
      kind: "err",
      why: `Directory doesn't contain a ${filename} file`
    };
  }
  const definitionStat = ctx.fs.stat(definitionPath);
  if (!definitionStat.isFile()) {
    return {
      kind: "err",
      why: `Component definition ${filename} isn't a file`
    };
  }
  return {
    kind: "ok",
    component: {
      isRoot,
      path: path.resolve(directory),
      definitionPath
    }
  };
}
export async function buildComponentDirectory(ctx, definitionPath) {
  const isRoot = path.basename(definitionPath) === ROOT_DEFINITION_FILENAME;
  const isComponent = isComponentDirectory(
    ctx,
    path.dirname(definitionPath),
    isRoot
  );
  if (isComponent.kind === "err") {
    logError(
      ctx,
      `Invalid component directory (${isComponent.why}): ${path.dirname(definitionPath)}`
    );
    return await ctx.crash(1, "invalid filesystem data");
  }
  return isComponent.component;
}
export function toComponentDefinitionPath(rootComponent, component) {
  return path.relative(
    rootComponent.path,
    component.path
  );
}
export function toAbsolutePath(rootComponent, componentDefinitionPath) {
  return path.normalize(path.join(rootComponent.path, componentDefinitionPath));
}
